### PR TITLE
FCS: Enable error reporting with correct file name

### PR DESCRIPTION
**Pull request by goswinr**
Currently, compiler errors in fsi evaluations via FCS always report an error in a file called `input.fsx`. 
This pull request adds an ~~optional parameter~~ _overload_ that allows specifying a filename related to an evaluated string.

For example, calling `fsiSession.EvalInteraction( "1 + 1.0" )` reports `input.fsx (1,3)-(1,4) typecheck error The type 'float' does not match the type 'int'`
But the actual file where this string comes from might be named differently.

I don't want to use `fsiSession.EvalScript(path)` since a file might not exist yet, or I might not be evaluating all lines of the current file.

### CODE REVIEW

**Comment by T-Gro** [comment]
I know it probably isn't a big deal, but I would prefer if this would be done without breaking binary compatibility.
i.e. adding a new overload, and keeping the existing one WITHOUT requiring consuming code to be recompiled.

(in this case the code overhead is negligable)

**Comment by goswinr** [comment]
@T-Gro sure , I can change the PR. I was not sure how to best expose the `dummyFileName`. So instead of two optional members there should be and overload with two required members ( and the cancellation token still optional) ?

**Comment by T-Gro** [comment]
> @T-Gro sure , I can change the PR. I was not sure how to best expose the `dummyScriptFileName `. So instead of two optional members there should be and overload with two required members ( and the cancellation token still optional) ?

Yep.

Because adding an optional member to existing functionality requires recompilation (with no code changes on consumer side), but is binary breaking.

A new overload would keep existing usages working.

File: src/Compiler/Interactive/fsi.fs [hunk 1]
```diff
@@ -3631,48 +3631,52 @@ type FsiEvaluationSession (fsi: FsiEvaluationSessionHostConfig, argv:string[], i
...
         // is not safe to call concurrently.
         let ctok = AssumeCompilationThreadWithoutEvidence()
         let cancellationToken = defaultArg cancellationToken CancellationToken.None
+        let scrptFileName = defaultArg scriptFileName dummyScriptFileName
 
         let errorOptions = TcConfig.Create(tcConfigB,validate = false).diagnosticsOptions
         let diagnosticsLogger = CompilationDiagnosticLogger("EvalInteraction", errorOptions, eagerFormat)
-        fsiInteractionProcessor.EvalInteraction(ctok, code, dummyScriptFileName, diagnosticsLogger, cancellationToken)
-        |> commitResultNonThrowing errorOptions "input.fsx" diagnosticsLogger
+        fsiInteractionProcessor.EvalInteraction(ctok, code, scrptFileName, diagnosticsLogger, cancellationToken)
```
Comment (KevinRansom) [inline src/Compiler/Interactive/fsi.fs:3678]: spelling : scrptFileName

**Review by KevinRansom — CHANGES_REQUESTED** [review]
As Tomas says, overloads are preferred for compatibility.

File: src/Compiler/Interactive/fsi.fs (hunk 1 above)
Comment (goswinr) [reply src/Compiler/Interactive/fsi.fs]: Thanks, this local value will be obsolete when moving to overloads.

**Comment by goswinr** [comment]
@T-Gro , @KevinRansom  I changed it to overloads. I also rebased it to use the new basline files for the API surface.

the new overloads are: 
```fsharp
member EvalInteraction: code: string * scriptFileName: string * ?cancellationToken: CancellationToken -> unit
member EvalInteractionNonThrowing: code: string * scriptFileName: string * ?cancellationToken: CancellationToken ->  Choice<FsiValue option, exn> * FSharpDiagnostic[]
member EvalExpression: code: string * scriptFileName: string -> FsiValue option
member EvalExpressionNonThrowing: code: string * scriptFileName: string -> Choice<FsiValue option, exn> * FSharpDiagnostic[]
```

the existing ones:
```fsharp
member EvalInteraction: code: string *  ?cancellationToken: CancellationToken -> unit
member EvalInteractionNonThrowing: code: string *  ?cancellationToken: CancellationToken ->  Choice<FsiValue option, exn> * FSharpDiagnostic[]
member EvalExpression: code: string  -> FsiValue option
member EvalExpressionNonThrowing: code: string  -> Choice<FsiValue option, exn> * FSharpDiagnostic[]
```

In my own tests the `scriptFileName` now also shows for `__SOURCE_FILE__` in scripts.

**Review by vzarytovskii — APPROVED** [review]

**Review by T-Gro — APPROVED** [review]

**Comment by goswinr** [comment]
Thanks @vzarytovskii ! Is there also some fcs documentation to update ?

**Comment by vzarytovskii** [comment]
> Thanks @vzarytovskii ! Is there also some fcs documentation to update ?

I don't think there's, docs should be automatically generated

**Comment by goswinr** [comment]
Thanks for merging @vzarytovskii ! However, I did not find this in the new Nuget that was published yesterday.
Is there a preview nuget that contains this change?
image

**Comment by vzarytovskii** [comment]
> Thanks for merging @vzarytovskii ! However, I did not find this in the new Nuget that was published yesterday.
> Is there a preview nuget that contains this change?
> image
> 

I don't think it made it to .200, it should be in the next .300 preview or our preview nuget feed.

**Pull request merged**